caddytest/caddytest.go

```diff
@@ -0,0 +1,224 @@
+package caddytest
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"path"
+	"runtime"
+	"strings"
+	"testing"
+	"time"
+)
+
+var internalHostnames = []string{"a.caddy.local", "b.caddy.local", "c.caddy.local"}
```

Comment:
Yeah that would be nice, I have had a look at this. I can override the dialer on the client which can override where it connects to _after_ DNS has been resolved. 

```
  {
    http_port     9080
    https_port    9443
  }
  
  www.cnn.com:9443 {
    tls /caddy.local.crt /caddy.local.key {
    }
    respond /version 200 {
      body "hello from a.caddy.local"
    }	
  }
```

for instance this works. 
```
caddytest.AssertGetResponse(t, "https://www.cnn.com:9443/version", 200, "hello from a.caddy.local")
```

But if the hostname is not successfully resolved then this process fails. We can force the DNS Resolver to call but it still needs a running DNS server to respond to it.

Will it be too confusing to have `www.cnn.com` in your config and not actually call that server? On the plus side if you do hardcode your server names in your config this is a pretty neat way to be able to test it.

Reply:
I was thinking there might be a way to override DNS resolution entirely -- like, dial a local address for everything, rather than using DNS at all.

Reply:
Won't providing customer `net.Dialer` to the [net.DefaultResolver](https://golang.org/pkg/net/#pkg-variables)? Of course with forcing Go-based DNS resolution with `GODEBUG=netdns=go`

Reply:
Thanks @Mohammed90 my read on the resolver is choose where to go to resolve the DNS, either head to `8.8.8.8` or `1.1.1.1`. You would need something listening on the other end to return a DNS response. Let me know if you think that is not the case.

Reply:
So checking out the end of this file https://golang.org/src/net/conf.go I thought I had something useful on fedora `nslookup something.localhost` will resolve to `127.0.0.1`

Reply:
@mholt I don't believe that there is a way to do this fully without a hosts file. It is the standard way - except on Windows?? for making this work. The windows server would not allow `ipconfig /flushdns' either which is odd, perhaps intentionally disabled for a build agent??

An interesting place to land is to force the dialer on to always connect to localhost by overriding the dialer. You can keep your hardcoded hostname `www.cnn.com` but it will test the config by dialing internally to 127.0.0.1. You will eventually hit limits like reverse proxying to another server in the same caddy config via a hostname, as this will be using caddys internal dialer etc.

I ruled out running a local dns server. Maybe caddy could host a dns server...

Reply:
https://coredns.io/ is actually built on top of Caddy v1, could run an instance of that?

Reply:
@sarge Yeah, this is kind of tricky.

Any way we can run the integration tests without relying on DNS names for now?

> Maybe caddy could host a dns server...

That's actually what https://coredns.io is. Just gotta see if it'll come to v2.

Reply:
A bit of a wild approach, but what if.... we run an instance of CoreDNS, set it as the system DNS server, configure it with all the local hosts for the tests, then kill it after we're done?

Reply:
(That's what I was trying to imply @Mohammed90 :+1:)

Reply:
Finding, installing and configuring coredns verses writing to the hosts file which is pretty standard (this windows build agent being an outlier), I would choose the hosts file. The tests do skip if the prerequisites are not met, so as to not impact people who are running without the hosts entries configured. Even if we are only testing on Linux and macOS I think there is merit in having this kind of testing. If windows support comes later then even better.

@mholt regarding not using dns names, I sure I can create some tests using `localhost` or `127.0.0.1` ip addresses.

Reply:
@sarge Here was one suggestion on Twitter: https://twitter.com/captncraig/status/1234639071307280385

What about that? Using http.Transport with a custom dialer.

Reply:
@mholt I have pushed an example of the custom dialer. Because DNS is resolved _before_ dialing it can help with a valid hardcoded DNS entry like `www.cnn.com` example. In the case of an non existent host like `randomhostname123.com`, when the DNS resolution fails, from the network stack there is reason to attempt the connection so it gives up before the dialer can be called.